controllers/cloudstackmachine_controller.go

```diff
@@ -268,21 +270,33 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateMachineStateChecker()
 }
 
 func (r *CloudStackMachineReconciliationRunner) ReconcileDelete() (retRes ctrl.Result, reterr error) {
-	if r.ReconciliationSubject.Spec.InstanceID != nil {
-		r.Log.Info("Deleting instance", "instance-id", r.ReconciliationSubject.Spec.InstanceID)
-		// Use CSClient instead of CSUser here to expunge as admin.
-		// The CloudStack-Go API does not return an error, but the VM won't delete with Expunge set if requested by
-		// non-domain admin user.
-		if err := r.CSClient.DestroyVMInstance(r.ReconciliationSubject); err != nil {
-			if err.Error() == "VM deletion in progress" {
-				r.Log.Info(err.Error())
-				return ctrl.Result{RequeueAfter: utils.DestoryVMRequeueInterval}, nil
-			}
+	if r.ReconciliationSubject.Spec.InstanceID == nil {
+		// InstanceID is not set until deploying VM finishes which can take minutes, and CloudStack Machine can be deleted before VM deployment complete.
+		// ResolveVMInstanceDetails can get InstanceID by CS machine name
+		err := r.CSClient.ResolveVMInstanceDetails(r.ReconciliationSubject)
+		if err != nil {
+			r.ReconciliationSubject.Status.Status = pointer.String(metav1.StatusFailure)
+			r.ReconciliationSubject.Status.Reason = pointer.String(err.Error() +
```

Comment:
do you have to defer a Patch operation on this object to get this status update to happen?

Reply:
Patch operation is part of the base reconciler which runs for all controller.